libsignal-protocol/src/crypto.rs

```diff
@@ -1,23 +1,291 @@
-use rand::RngCore;
+#[cfg(feature = "crypto-openssl")]
+use std::sync::{Arc, Mutex};
 use std::{
+    convert::TryFrom,
     os::raw::{c_int, c_void},
     pin::Pin,
     slice,
 };
+
 use sys::{signal_buffer, signal_crypto_provider};
 
+use crate::{
+    buffer::Buffer,
+    errors::{InternalError, IntoInternalErrorCode},
+};
+
+#[derive(Debug, Clone)]
+pub struct SignalCipherTypeError(i32);
+
+#[derive(Copy, Clone)]
+pub enum CipherMode {
+    Encrypt,
+    Decrypt,
+}
+pub enum SignalCipherType {
+    AesCtrNoPadding,
+    AesCbcPkcs5,
+}
+
+impl TryFrom<i32> for SignalCipherType {
+    type Error = SignalCipherTypeError;
+
+    #[inline]
+    fn try_from(v: i32) -> Result<Self, Self::Error> {
+        match v as u32 {
+            sys::SG_CIPHER_AES_CTR_NOPADDING => {
+                Ok(SignalCipherType::AesCtrNoPadding)
+            },
+            sys::SG_CIPHER_AES_CBC_PKCS5 => Ok(SignalCipherType::AesCbcPkcs5),
+            _ => Err(SignalCipherTypeError(v)),
+        }
+    }
+}
+
 /// Cryptography routines used in the signal protocol.
 pub trait Crypto {
-    fn fill_random(&self, buffer: &mut [u8]);
+    fn fill_random(&self, buffer: &mut [u8]) -> Result<(), InternalError>;
+    fn hmac_sha256_init(&self, key: &[u8]) -> Result<(), InternalError>;
+    fn hmac_sha256_update(&self, data: &[u8]) -> Result<(), InternalError>;
+    fn hmac_sha256_final(&self) -> Result<Vec<u8>, InternalError>;
+    fn hmac_sha256_cleanup(&self) {}
+
+    fn sha512_digest_init(&self) -> Result<(), InternalError>;
+    fn sha512_digest_update(&self, data: &[u8]) -> Result<(), InternalError>;
+    fn sha512_digest_final(&self) -> Result<Vec<u8>, InternalError>;
+    fn sha512_digest_cleanup(&self) {}
+
+    fn encrypt(
+        &self,
+        cipher: SignalCipherType,
+        key: &[u8],
+        iv: &[u8],
+        data: &[u8],
+    ) -> Result<Vec<u8>, InternalError>;
+    fn decrypt(
+        &self,
+        cipher: SignalCipherType,
+        key: &[u8],
+        iv: &[u8],
+        data: &[u8],
+    ) -> Result<Vec<u8>, InternalError>;
 }
 
+#[cfg(feature = "crypto-native")]
 #[derive(Debug, Copy, Clone, PartialEq)]
 pub struct DefaultCrypto;
 
+#[cfg(feature = "crypto-openssl")]
+pub struct OpenSSLCrypto {
+    hmac_ctx: Arc<Mutex<Option<openssl::hash::Hasher>>>,
```

Comment:
Looking through the code, I don't think these `Arc`'s are necessary. The `Crypto` object will already be behind a shared pointer (in this case c has a raw pointer to our pinned `Box<dyn Crypto>`) and I don't think we ever need to share the `hmac_ctx` or `sha512_ctx` with any other code.